File: haskell-code-mike/Contract.hs
module Contract where

{-
1. einfaches Beispiel
Zero-Bond / Zero Coupon Bond
"Ich bekomme am 24.12.2022 100€."

2. Beispiel zerlegen in "atomare Bestandteile" / "Ideen"
   (anfangen mit einzelnen Attributen)

- Währung ("Ich bekomme 1€ jetzt!")
- Betrag ("Ich bekomme jetzt 100€")
- Später

Currency swap:
   Am 24.12.2022:
   - ich bekomme 100€
   - ich bezahle 150GBP
-}

data Date = MkDate String deriving (Eq, Ord, Show)

type Amount = Float

data Currency = EUR | GBP | USD | YEN
  deriving Show
{-
data Contract =
    ZeroCouponBond Date Amount Currency
  | Future Date -- ...
  | Put
  | Call
    deriving Show

zcb1 :: Contract
zcb1 = ZeroCouponBond (MkDate "2022-12-24") 100 EUR
-}

data Contract =
    One Currency
  | Many Amount Contract  -- <<< Selbstbezug
  | Later Date Contract
  | Give Contract
  | And Contract Contract -- Halbgruppe
  | Zero -- neutrale Element
  deriving Show

c1 = One EUR -- "Ich bekomme 1€ jetzt."
c2 = Many 100 (One EUR) -- "Ich bekomme 100€ jetzt."

c3 = Many 10 (Many 20 (One EUR)) -- 200€

zcb1 = Later (MkDate "2022-12-24") (Many 100 (One EUR))

zeroCouponBond :: Date -> Amount -> Currency -> Contract
zeroCouponBond date amount currency =
    Later date (Many amount (One currency))

zcb1' = zeroCouponBond (MkDate "2022-12-24") 100 EUR

zcb2 = Give (zeroCouponBond (MkDate "2022-12-24") 150 GBP)

currencySwap date amount1 currency1 amount2 currency2 =
    And (zeroCouponBond date amount1 currency1)
        (Give (zeroCouponBond date amount2 currency2))

-- im Paper:
-- scale :: Obs Double -> Contract -> Contract

-- haben Syntax
-- brauchen Semantik

File: scala-mike/src/main/scala/funar/Contract.scala
package funar

import javax.swing.plaf.BorderUIResource.EmptyBorderUIResource

/*
Finanzderivate:

Zins-Swap
in der Zukunft:
bekomme festen Zinssatz
zahle variablen Zinssatz

2 Vertragspartner:innen

1. finde ein einfaches Beispiel

Rollover-Geschäft: heute was bekommen, morgen was mit Zinsen zahlen

- heute was bekommen
- morgen was mit Zinsen zahlen
- morgen was (ohne Zinsen) zahlen
- morgen was (ohne Zinsen) bekommen

"Am 24.12. bekomme ich 100€": Zero-Bond, Zero-Coupon Bond

2. einfache Beispiel in "atomare Bestandteile" zerlegt

3 Ideen:

- € "Währung"
- Vielfaches
- Später

3. finde Selbstbezüge / Kombinatoren

4. sukzessive größere Beispiele abbilden, ggf. neue Kombinatoren definieren
*/

case class Date(iso: String) {
  def before(other: Date): Boolean = this.iso < other.iso
}

type Amount = Double

enum Currency {
  case EUR 
  case GBP
  case USD
}

/*
enum Contract {
  case ZeroCouponBond(date: Date, amount: Amount, currency: Currency)
}
*/

enum Contract {
  case One(currency: Currency)
  // case Multiple(amount: Amount, currency: Currency)
  case Multiple(amount: Amount, contract: Contract)
  case Forward(date: Date, contract: Contract)
  case Reverse(contract: Contract)
  case Combined(contract1: Contract, contract2: Contract)
  case Zero
}

object Contract {
  import Currency._
  import Contract._
  val zcb1 = Forward(Date("2022-12-24"), Multiple(100, One(EUR)))

  def zeroCouponBond(date: Date, amount: Amount, currency: Currency): Contract =
    Forward(date, Multiple(amount, One(currency)))

  val zcb1_ = zeroCouponBond(Date("2022-12-24"), 100, EUR)

  val rollover = Combined(zeroCouponBond(Date("2022-03-29"), 100, EUR),
                          Reverse(zeroCouponBond(Date("2022-03-30"), 105, EUR)))


  enum Direction {
    case Long 
    case Short

    def invert = this match {
      case Long => Short
      case Short => Long
    }
  }

  case class Payment(date: Date, direction: Direction, amount: Amount, currency: Currency) {
    def invert = this.copy(direction = this.direction.invert)
    def scale(factor: Double) = this.copy(amount = this.amount * factor)
  }

  // alle Zahlungen, die bis heute fällig sind
  // heraus kommt ein "Residualvertrag"
  def meaning(contract: Contract, today: Date): (Seq[Payment], Contract) = ???
}